### PR TITLE
Fixes bad escaping of String values in generated migration code

### PR DESCRIPTION
**Pull request by CZEMacLeod**
To address issue #198
String values containing quotes or other language specific strings are not escaped properly when generating migration code.
This change replaces the escaping code with System.CodeDom.Compiler.CodeDomProvider to allow generating correctly escaped strings.

### CODE REVIEW

**Comment by rowanmiller** [comment]
We're not sure that this is the right fix. It seems like the purpose of this method is specifically to quote identifiers. Maybe the code that writes out the string values should be calling a separate method to escape them. Assigning to @bricelam to investigate.

File: src/EntityFramework/Migrations/Design/CSharpMigrationCodeGenerator.cs [hunk 1]
```diff
@@ -1536,7 +1536,14 @@ protected virtual string TranslateColumnType(PrimitiveTypeKind primitiveTypeKind
...
         {
-            return "\"" + identifier + "\"";
+            using (var writer = new StringWriter())
+            {
+                using (var provider = CodeDom.Compiler.CodeDomProvider.CreateProvider("CSharp"))
+                {
+                    provider.GenerateCodeFromExpression(new CodeDom.CodePrimitiveExpression(identifier), writer, null);
+                    return writer.ToString();
+                }
+            }
```
Comment (bricelam) [inline src/EntityFramework/Migrations/Design/CSharpMigrationCodeGenerator.cs]: I much prefer you're original suggestion to using CodeDOM. The following is probably sufficient. (It's what we use in EF Core)
```C#
return "\"" + identifier.Replace("\\", "\\\\").Replace("\"", "\\\"") + "\"";
```

File: src/EntityFramework/Migrations/Design/VisualBasicMigrationCodeGenerator.cs [hunk 2]
```diff
@@ -1577,7 +1577,14 @@ protected virtual string TranslateColumnType(PrimitiveTypeKind primitiveTypeKind
...
         {
-            return "\"" + identifier + "\"";
+            using (var writer = new StringWriter())
+            {
+                using (var provider = CodeDom.Compiler.CodeDomProvider.CreateProvider("VisualBasic"))
+                {
+                    provider.GenerateCodeFromExpression(new CodeDom.CodePrimitiveExpression(identifier), writer, null);
+                    return writer.ToString();
+                }
+            }
```
Comment (bricelam) [inline src/EntityFramework/Migrations/Design/VisualBasicMigrationCodeGenerator.cs]: For VB, I think this is sufficient:
```C#
return "\"" + identifier.Replace("\"", "\"\"") + "\"";
```

**Comment by bricelam** [comment]
@rowanmiller it's the right method. `identifier` is a poor parameter name, and it was probably copied from something generating SQL...

File: src/EntityFramework/Migrations/Design/CSharpMigrationCodeGenerator.cs (hunk 1 above)
Comment (rowanmiller) [inline src/EntityFramework/Migrations/Design/CSharpMigrationCodeGenerator.cs]: We discussed that in triage today too... we couldn't remember if we used CodeDOM already or not.

Comment (CZEMacLeod) [inline src/EntityFramework/Migrations/Design/CSharpMigrationCodeGenerator.cs]: Sorry for the delay - I have been using my version including CodeDOM for a while now and in general I think the CodeDOM approach works better when generating long strings etc. as it adds line breaks as required.
However, I did find an issue when generating migrations in a class library with a startup project which included the Microsoft.CodeDom.Providers.DotNetCompilerPlatform package as it resulted in not finding the correct version of the providers unless the package was also added to the class library. Although this did result in using the latest Roslyn based providers.
I am happy to switch to the 'simpler' Replace approach for the providers as suggested and will update my pull appropriately.

**Comment by bricelam** [comment]
I've reverted this. We had some CI failures that may be related. I've resubmitted the changes in #232 so we can try again before the 6.3.0 release.

**Comment by ErikEJ** [comment]
@bricelam  What is the release roadmap for EF 6.2 / EF 6.3 actually?

**Comment by bricelam** [comment]
lol, no idea

**Comment by divega** [comment]
@erikej The main issue we were waiting to resolve was the story for shipping a beta version of the EF6 Designer that could be downloaded out of band and installed on VS 2017. We have sorted that out (we won't do a beta for VS2017 but we will use the preview channel at a later point), so now it is a matter of making spending a few cycles to put everything together. There is also one bug that it is still a candidate to be fixed in the beta: #218. Hopefully next week we will be able to pick a date.

**Pull request merged**